server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java

```diff
@@ -119,61 +134,114 @@ public void exit() {
   enum Status {
     UNINITIALIZED,
     INITIALIZED,
-    SHUTDOWN;
+    SHUTDOWN
   }
 
   /**
-   * Try to update the Gradle Java home if:
-   * <ul>
-   *   <li>Gradle Java home is not set.</li>
-   *   <li>A valid JDK can be found to launch Gradle.</li>
-   * </ul>
-   *
-   * <p>The JDK installation path string will be set to {@link Preferences#gradleJavaHome}.
+   * Finds a suitable version of JDK to use for compilation of Java programs.
+   * Returns {@code null} if no suitable JDK can be found.
    */
-  private void updateGradleJavaHomeIfNecessary(URI rootUri) {
+  private File getSuitableJdk(URI rootUri) {
+
     Preferences preferences = preferenceManager.getPreferences();
-    if (preferences.getJdks() == null || preferences.getJdks().isEmpty()) {
-      return;
+    String filePath = preferences.getGradleJavaHome();
+    File gradleJavaHome;
+    if (StringUtils.isNotEmpty(filePath)) {
+      gradleJavaHome = new File(filePath);
+    } else {
+      gradleJavaHome = connector.getGradleJavaHome(rootUri);
     }
 
-    if (StringUtils.isBlank(preferences.getGradleJavaHome())) {
-      String gradleVersion = "";
-      GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
-      Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
-      LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
-      if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
-        gradleVersion = preferences.getGradleVersion();
-      } else {
-        gradleVersion = connector.getGradleVersion(rootUri);
-      }
+    GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
+    Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
+    LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
+
+    String gradleVersion;
+    if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
+      gradleVersion = preferences.getGradleVersion();
+    } else {
+      gradleVersion = connector.getGradleVersion(rootUri);
+    }
+
+    if (StringUtils.isEmpty(gradleVersion)) {
+      return null;
```

Comment:
Do we know when `gradleVersion` will be null?

Reply:
I think you meant empty? The answer is yes, when retrieving the Gradle version using `GradleApiConnector#getGradleVersion` from the BuildEnvironment and a BuildException gets thrown. In this case we will get an empty string.

Reply:
Do you know when will `BuildEnvironment` get thrown? Just asking for curiosity.

Reply:
The check already existed in the codebase. Although [BuildExceptions](https://docs.gradle.org/current/javadoc/org/gradle/tooling/BuildException.html) are thrown when a model cannot be built. Based on [this](https://stackoverflow.com/questions/53991804/could-not-fetch-model-of-type-buildenvironment-using-gradle-distribution) I would say building the `BuildEnvironment` model would throw an exception due to incorrectly defined Gradle properties.